rustworkx-core/src/dag_algo.rs

```diff
@@ -313,6 +314,161 @@ where
     Ok(Some((path, path_weight)))
 }
 
+/// Collect runs that match a filter function given edge colors.
+///
+/// A bicolor run is a list of groups of nodes connected by edges of exactly
+/// two colors. In addition, all nodes in the group must match the given
+/// condition. Each node in the graph can appear in only a single group
+/// in the bicolor run.
+///
+/// # Arguments:
+///
+/// * `dag`: The DAG to find bicolor runs in
+/// * `filter_fn`: The filter function to use for matching nodes. It takes
+///     in one argument, the node data payload/weight object, and will return a
+///     boolean whether the node matches the conditions or not.
+///     If it returns ``true``, it will continue the bicolor chain.
+///     If it returns ``false``, it will stop the bicolor chain.
+///     If it returns ``None`` it will skip that node.
+/// * `color_fn`: The function that gives the color of the edge. It takes
+///     in one argument, the edge data payload/weight object, and will
+///     return a non-negative integer, the edge color. If the color is None,
+///     the edge is ignored.
+///
+/// # Returns:
+///
+/// * `Vec<Vec<G::NodeId>>`: a list of groups with exactly two edge colors, where each group
+///     is a list of node data payload/weight for the nodes in the bicolor run
+/// * `None` if a cycle is found in the graph
+/// * Raises an error if found computing the bicolor runs
+///
+/// # Example:
+///
+/// ```rust
+/// use rustworkx_core::dag_algo::collect_bicolor_runs;
+/// use petgraph::graph::DiGraph;
+/// use std::error::Error;
+///
+/// let mut graph = DiGraph::new();
+/// let n0 = graph.add_node(0);
+/// let n1 = graph.add_node(1);
+/// let n2 = graph.add_node(2);
+/// let n3 = graph.add_node(3);
+/// graph.add_edge(n0, n2, 0);
+/// graph.add_edge(n1, n2, 0);
+/// graph.add_edge(n2, n3, 0);
+/// graph.add_edge(n2, n3, 0);
+/// // filter_fn and color_fn must share error type
+/// fn filter_fn(node: &i32) -> Result<Option<bool>, Box<dyn Error>> {
+///  Ok(Some(*node > 1))
+/// }
+/// fn color_fn(edge: &i32) -> Result<Option<usize>, Box<dyn Error>> {
+///   Ok(Some(*edge as usize))
+/// }
+/// let result = collect_bicolor_runs(&graph, filter_fn, color_fn).unwrap();
+/// ```
+pub fn collect_bicolor_runs<G, F, C, E>(
+    graph: G,
+    filter_fn: F,
+    color_fn: C,
+) -> Result<Option<Vec<Vec<G::NodeId>>>, E>
+where
+    F: Fn(&<G as Data>::NodeWeight) -> Result<Option<bool>, E>,
+    C: Fn(&<G as Data>::EdgeWeight) -> Result<Option<usize>, E>,
+    G: IntoNodeIdentifiers // Used in toposort
+        + IntoNeighborsDirected // Used in toposort
+        + IntoEdgesDirected // Used for .edges_directed
+        + Visitable // Used in toposort
+        + DataMap, // Used for .node_weight
```

Comment:
Hmm I agree that this would make the code more readable, and more general, as the filtering/coloring functions could also just depend on the node/edge id. At the same time, I don't know if many coloring functions would use the ids directly, they would probably access the weights (especially with larger graphs, handling 30 node ids vs, let's say, 4 weights, looks painful), so we would be adding that "extra work" to the users. However, I think accessing the weights from the ids is not too bad or surprising, so we can go the general route. I'll apply the changes and see if I see anything else.

Reply:
FWIW, the only thing I know using this function is Qiskit, we added it to rustworkx based on an algorithm we were doing in Python previously: https://github.com/Qiskit/rustworkx/issues/265. It's kind of a bespoke thing we wanted to do for Qiskit's dag traversal. The reason we're porting it is to rustworkx-core is so we can call it from Rust in Qiskit :) 

So if using weights here makes the Qiskit use case easier I guess we can leave it as is, but with my rustworkx-core api design hat on I felt using ids felt more natural.